sdk/lib/opentelemetry/sdk/trace/span.rb

```diff
@@ -343,19 +343,18 @@ def trim_span_attributes(attrs)
 
           excess = attrs.size - @span_limits.attribute_count_limit
           excess.times { attrs.shift } if excess.positive?
-          truncate_attribute_values(attrs)
+          truncate_attribute_values(attrs, @span_limits.span_attribute_length_limit)
           nil
         end
 
-        def truncate_attribute_values(attrs)
+        def truncate_attribute_values(attrs, attribute_length_limit)
           return EMPTY_ATTRIBUTES if attrs.nil?
 
-          attribute_length_limit = @span_limits.attribute_length_limit
           attrs.each { |key, value| attrs[key] = OpenTelemetry::Common::Utilities.truncate(value, attribute_length_limit) } if attribute_length_limit
           attrs
         end
 
-        def trim_links(links, link_count_limit, link_attribute_count_limit) # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+        def trim_links(links, link_count_limit, link_attribute_count_limit, attribute_length_limit) # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
           # Fast path (likely) common cases.
           return nil if links.nil?
 
```

Comment:
Ah, thanks, I see now. But hm, should `valid_attributes?` check the size and return false if a value is too large? That may have implications in other parts of the code, but the other option would be to do another iteration over the attributes.

Reply:
This seems inefficient, but all I've come up with so far:

```ruby
          if links.size <= link_count_limit &&
             links.all? { |link| link.span_context.valid? && link.attributes.size <= link_attribute_count_limit && Internal.valid_attributes?(name, 'link', link.attributes) }
            links.map! do |link|
              attrs = Hash[link.attributes] # link.attributes is frozen, so we need an unfrozen copy to adjust.
              OpenTelemetry::Trace::Link.new(link.span_context, truncate_attribute_values(attrs, attribute_length_limit))
            end
            return links.frozen? ? links : links.clone.freeze
          end
```

Reply:
the idea is to provide a fast path for "no changes required". Since we're adding an `attribute_length_limit`, I think we should pass that to `valid_attributes?` and it should return `false` if any attribute exceeds the limit.

There will be a point at which this "fast path" is fast enough to justify the complexity. We haven't benchmarked it - this is largely based on intuition that most of the time instrumentation won't hit any of the limits.